fwupd: 1.7.2 -> 1.7.6

tomfitzhenry opened the pull request:
###### Motivation for this change

- https://github.com/fwupd/fwupd/releases/tag/1.7.3
- https://github.com/fwupd/fwupd/releases/tag/1.7.4
- https://github.com/fwupd/fwupd/releases/tag/1.7.5
- https://github.com/fwupd/fwupd/releases/tag/1.7.6

Update fwupd to 1.7.6 to support Pinephone modem firmware upgrades, as described in https://gitlab.com/linux-mobile/tracker/-/issues/11 and https://gitlab.com/postmarketOS/pmaports/-/merge_requests/2760.

I had to update `add-option-for-installation-sysconfdir.patch`, since the patch no longer successfully applied.

This worked (caveat: https://github.com/NixOS/mobile-nixos/issues/463), and I'm now running the https://github.com/Biktorgj/pinephone_modem_sdk firmware successfully.

`fwupd.tests` is now passing.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

tomfitzhenry committed:
fwupd: 1.7.2 -> 1.7.6

https://github.com/fwupd/fwupd/releases/tag/1.7.3
https://github.com/fwupd/fwupd/releases/tag/1.7.4
https://github.com/fwupd/fwupd/releases/tag/1.7.5
https://github.com/fwupd/fwupd/releases/tag/1.7.6

jtojnar commented on pkgs/os-specific/linux/firmware/fwupd/default.nix:
```diff
@@ -50,6 +50,9 @@
+, libmbim
 }:
 
 let
```
Looks like `fwupd-efi` and `fwupd.test-firmware` are up to date.

tomfitzhenry committed:
fwupd: enable modemmanager plugin

tomfitzhenry commented on pkgs/os-specific/linux/firmware/fwupd/default.nix:
```diff
@@ -112,16 +115,18 @@ let
-    src = fetchurl {
-      url = "https://people.freedesktop.org/~hughsient/releases/fwupd-${version}.tar.xz";
-      sha256 = "sha256-hjLfacO6/Fk4fNy1F8POMaWXoJAm5E9ZB9g4RnG5+DQ=";
+    src = fetchFromGitHub {
```
> Why switch away from tarballl?

Github is the upstream VCS for fwupd (per https://fwupd.org/lvfs/docs/developers ). I think using the VCS repo rather than a tarball makes it easier for NixOS users/developers to find the upstream code.

More likely that bots will be able to propose PRs for new releases.

I'm happy to switch back to the tarball though.

> Also the hash is wrong.

Oops, sorry. While developing this PR, I had initially used https://github.com/fwupd/fwupd/commit/2cd58847fc456de0c6a2270e67fea2842f3c07cb as the `rev`, and set the hash to be the hash of that. When I eventually set `rev` to be a tag, I forgot to update the hash, and the Nix build cache was happy to use its cached version (rather than download the tag's copy, and alert me that the hash was different).

Fixed.

jtojnar replied:
Generally, we prefer using the official tarballs when available, since that is most likely what is tested by upstream.

tomfitzhenry replied:
SGTM. Fixed.

tomfitzhenry commented:
> Looks like more changes are needed according to `fwupd.tests`:

Done. I've added `fwupd/msr.conf` to `filesInstalledToEtc`.

`nix-build -A fwupd.tests` continues to fail for other reason, however:

https://pastebin.com/9bCZ741A

Excerpt:
```
Running test: fwupd/fwupdmgr-p2p.test
Starting P2P daemon...
/nix/store/7mhpzl751n6056vjxb87z3ry1w0g6b2r-fwupd-1.7.6-installedTests/share/installed-tests/fwupd/fwupdmgr-p2p.sh: line 13: /usr/libexec/fwupd/fwupd: No such file or directory
Executing: fwupd/fwupdmgr-p2p.test
Executing: fwupd/fwupdmgr-p2p.test
Executing: fwupd/fwupdmgr-p2p.test
```

Sadly, `fwupd` is not installed to `bin`, so we'll have to replace fwupdmgr-p2p.sh: in nix.

tomfitzhenry commented:
> Sadly, fwupd is not installed to bin, so we'll have to replace fwupdmgr-p2p.sh: in nix.

Done. This `fwupd.tests` is now passing.

jtojnar commented on pkgs/os-specific/linux/firmware/fwupd/default.nix:
```diff
@@ -257,6 +258,10 @@ let
         po/test-deps
+
+      substituteInPlace data/installed-tests/fwupdmgr-p2p.sh \
+        --replace "/usr/libexec/fwupd/fwupd" $out/libexec/fwupd/fwupd \
```
The FHS path should be fixed upstream. Could you please try to make a PR?

tomfitzhenry replied:
What should it be?

/usr/libexec/fwupd/fwupd seems to comply with https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s07.html

jtojnar replied:
It should be `@libexecdir@/fwupd/fwupd` and substituted using Mesonʼs `configure_file` function.

tomfitzhenry replied:
Ah, that makes sense. I'm not familiar with Meson so have raised this as https://github.com/fwupd/fwupd/issues/4360.

jtojnar committed:
fwupd: use upstream install detection patch

https://github.com/fwupd/fwupd/commit/5bc546221331feae9cedc1892219a25d8837955f
attempts to fix the need to patch the redfish test ourselves.
But it is actually not sufficient since we install the installed tests
into a separate prefix, so we need to adjust the installed tests patch.

jtojnar reviewed (approved):
Picked the upstream patch for the FHS path and also for the redfish tests. Looks good to me now.
